### DISCUSSION TITLE
Instantiation syntax

### DISCUSSION THREAD

**Discussion by cmnrd** [lf-lang/lingua-franca#2044] [Ideas]
I think that there is a severe inconsistency in our syntax. Most of the elements in a reactor look like this:
```
state foo: int = 42
input in: void
output out: void
logical action a: void
```

This could be summarized using the following rule:
```
// <keyword> <name> (: <type>)? (= <expr>)? ((<expr>*))?
```

However, the syntax for reactor instances uses a completely different scheme:
```
bar = new Bar(bar=42) 
```

Besides the inconsistency, there is another problem with using the `new` keyword. `new` has a very specific meaning in various programming languages (dynamic memory allocation). However, that is not what is happening in several of our implementations. I think that using `new` in LF is very misleading, since LF does not do any memory management.

Instead, we could use this syntax:
```
instance bar = Bar(bar=42)
```

We could also allow an optional type annotation, in which case all the following would be valid (this is inspired by Kotlin):
```
instance bar: Bar = Bar(bar=42)  // Bar initialized with bar=42
instance bar = Bar(bar=42) // type is trivially inferred from the assignment
instance bar: Bar // not initialized, use default constructor
```

What do you think?

**Comment by edwardalee** [comment]
Yikes.  Don't we have higher priorities?

**Reply by cmnrd** [reply]
This might seem unimportant (and I agree that this is not urgent), but I think we should aim at stabilizing our syntax and removing the quirks that it still has before attracting a wider audience. I am bringing this up because I am putting the core syntax rules of LF in my diss (and maybe also in the upcoming TODAES paper if we go for it), and I would like this to be as final as possible.

Consistency is important when designing a language. I have stumbled over the instance syntax several times, and I am a seasoned LF programmer. But if I am the only one who thinks this is a flaw in our language, I am happy to drop this. Most importantly, I would like to understand if others also see this as a problem.

Implementing the proposed syntax changes is straight forward (just edit the xtext grammar), and we can use the formatter to automatically convert old LF code. So the biggest chunk of work would be updating the docs.

**Comment by marcusrossel** [comment]
Sidenote: `instance` is also a keyword in languages with type classes like Haskell and Lean, though I'm guessing `new` is much more common - especially in languages which are interesting in the context of CPS.

**Comment by lhstrh** [comment]
I see your point, but I have to admit that the syntax for creating instances hasn't bothered me personally. That being said, a more consistent syntax does appeal to me. It would be nice to get input from more people on this... Let's discuss it during our weekly meeting?
